### PR TITLE
feat: mm linked pools

### PR DESCRIPTION
**Pull request by Chef-Yogi**

### CODE REVIEW

File: apps/web/src/hooks/useContract.ts [hunk 1]
```diff
@@ -384,3 +385,9 @@ export const useCrossFarmingProxy = (proxyContractAddress: string, withSignerIfP
     [proxyContractAddress, providerOrSigner, chainId],
   )
 }
+
+export const useMMLinkedPoolContract = () => {
```
Comment (chef-jojo) [inline apps/web/src/hooks/useContract.ts:389]: not in used?

File: apps/web/src/views/Swap/MMLinkPools/components/MMCommitButton.tsx [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  recipient,
+  allowedSlippage,
+  parsedIndepentFieldAmount,
+  onUserInput,
+  rfq,
+  refreshRFQ,
+  isRFQLoading,
+}: SwapCommitButtonPropsType) {
+  const { t } = useTranslation()
+  const [singleHopOnly] = useUserSingleHopOnly()
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/components/MMCommitButton.tsx:89]: what is single hop on MM?

File: apps/web/src/views/Swap/SmartSwap/index.tsx [hunk 3]
```diff
@@ -99,8 +103,22 @@ export const SmartSwapForm: React.FC<{ handleOutputSelect: (newCurrencyOutput: C
     inputError: stableSwapInputError,
   } = useDerivedSwapInfoWithStableSwap(independentField, typedValue, inputCurrency, outputCurrency, recipient)
 
+  const mmOrderBookTrade = useMMTrade(independentField, typedValue, inputCurrency, outputCurrency, recipient)
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/SmartSwap/index.tsx:106]: remove `recipient`

File: apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    // eslint-disable-next-line no-unsafe-finally
+    return parseAmountString
+  }
+}
+
+export const parseMMParameter = (
+  chainId: number,
+  inputCurrency: Currency,
+  outputCurrency: Currency,
+  independentField: Field,
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts:111]: be aware of optional

**Comment by chef-jojo** [comment]
did you add support chain id to skip this feature in other chains?

File: apps/web/src/views/Swap/MMLinkPools/hooks/useIsMMTradeBetter.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+export const useIsTradeWithMMBetter = ({ trade, v2Trade, tradeWithMM }: Options) => {
+  if (!tradeWithMM || tradeWithMM.inputAmount.equalTo(ZERO) || tradeWithMM.outputAmount.equalTo(ZERO)) {
+    return false
+  }
+  if (!v2Trade && !trade && tradeWithMM) return true // v2 and smart router has not liq but MM provide the deal
+  if (!v2Trade?.outputAmount || !v2Trade?.inputAmount) {
+    if (tradeWithMM) return true
+  }
+  return (
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/hooks/useIsMMTradeBetter.tsx:19]: It is not a hook overall

File: apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+import { OrderBookRequest, TradeWithMM } from '../types'
+
+const NATIVE_CURRENCY_ADDRESS = getAddress('0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE')
+
+export function useMMSwapContract() {
+  const { chainId } = useActiveChainId()
+  return useContract<MmLinkedPool>(MM_SWAP_CONTRACT_ADDRESS[chainId], PancakeSwapMMLinkedPoolABI, true)
+}
+
+export function calculateSlippageAmount(value: CurrencyAmount<Currency>, slippage: number): [JSBI, JSBI] {
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts:23]: remove

File: apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  return (
+    trade.route.path.some((token) => token.isToken && token.address === checksummedAddress) ||
+    trade.route.pairs.some((pair) => (pair as Pair)?.liquidityToken?.address === checksummedAddress)
+  )
+}
+
+export const useOrderBookQuote = (request: OrderBookRequest): OrderBookResponse => {
+  const { data } = useSWR(
+    request &&
+      request.trader &&
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts:32]: should fetch without connect wallet?

Comment (chef-jojo) [reply apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts:32]: block fetching when is native

File: apps/web/src/views/Swap/MMLinkPools/apis/index.ts [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  try {
+    const response = await fetch(`${API_ENDPOINT}/order-book-price`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(param),
+    })
+    const data = await response.json()
+    return data as OrderBookResponse
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/apis/index.ts:15]: use zod when you have time to refactor

File: apps/web/src/hooks/useContract.ts (hunk 1 above)
Comment (Chef-Yogi) [reply apps/web/src/hooks/useContract.ts:389]: removed

File: apps/web/src/views/Swap/MMLinkPools/hooks/useIsMMTradeBetter.tsx (hunk 5 above)
Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/hooks/useIsMMTradeBetter.tsx]: updated

File: apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts (hunk 4 above)
Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts]: done

**Comment by Chef-Yogi** [comment]
> did you add support chain id to skip this feature in other chains?

yeah have a hook for that `useIsMMSupportChain`

File: apps/web/src/views/Swap/MMLinkPools/hooks/useSwapCallback.ts [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+              },
+              type: 'swap',
+            })
+            logSwap({
+              chainId,
+              inputAmount,
+              outputAmount,
+              input: trade.inputAmount.currency,
+              output: trade.outputAmount.currency,
+              type: 'SmartSwap',
```
[the remark this answers is not in the document]
Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/hooks/useSwapCallback.ts:174]: Done

File: apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts (hunk 6 above)
Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/utils/exchange.ts]: Done

File: apps/web/next.config.mjs [hunk 10]
```diff
@@ -74,6 +74,10 @@ const config = {
         source: '/info/pool/:address',
         destination: '/info/pools/:address',
       },
+      {
+        source: '/mmLinkedPool/:path*',
+        destination: 'https://test.linked-pool.pancakeswap.com/quote-service/:path*',
+      },
```
Comment (chef-jojo) [inline apps/web/next.config.mjs:80]: still getting CORS?

File: apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  currencies: { [field in Field]?: Currency }
+  currencyBalances: { [field in Field]?: CurrencyAmount<Currency> }
+  parsedAmount: CurrencyAmount<Currency> | undefined
+  trade?: TradeWithMM<Currency, Currency, TradeType> | null
+  inputError?: string
+  mmParam: OrderBookRequest
+} | null => {
+  const { account, chainId } = useActiveWeb3React()
+  const mmParam = useMemo(
+    () => parseMMParameter(chainId, inputCurrency, outputCurrency, independentField, typedValue, account),
```
[the remark this answers is not in the document]
Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts:61]: done

File: apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts (hunk 7 above)
Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/hooks/useMMOrderBookTrade.ts:32]: Done

File: apps/web/next.config.mjs (hunk 10 above)
Comment (Chef-Yogi) [reply apps/web/next.config.mjs:80]: for now, yes, already ask to BE to add  Access-Control-Allow-Origin on the header, then we can remove this

Comment (Chef-Yogi) [reply apps/web/next.config.mjs]: done

File: apps/web/src/views/Swap/MMLinkPools/constants.ts [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import { ChainId } from '@pancakeswap/sdk'
+
+export const MM_SUPPORT_CHIAN = {
+  1: true,
+  5: true,
+  56: false,
+}
+
+export const MM_SWAP_CONTRACT_ADDRESS = {
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/MMLinkPools/constants.ts:9]: Same in contract.ts?

Comment (Chef-Yogi) [reply apps/web/src/views/Swap/MMLinkPools/constants.ts]: Done

**Review by chef-jojo — APPROVED** [review]

**Pull request merged**